Mathlib/Tactic/Simp/FinsetInterval.lean

```diff
@@ -0,0 +1,140 @@
+/-
+Copyright (c) 2025 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import Mathlib.Order.Interval.Finset.Nat
+import Mathlib.Util.Qq
+import Mathlib.Tactic.NormNum.Basic
+
+/-!
+# Simproc for intervals of natural numbers
+-/
+
+open Qq Lean Mathlib.Meta.NormNum Finset
+
+namespace Mathlib.Tactic.Simp
+section lemmas
+variable {m n : ℕ} {s : Finset ℕ}
+
+private lemma Icc_eq_insert_of_Icc_succ_eq (hmn : m ≤ n) (hs : Icc (m + 1) n = s) :
+    Icc m n = insert m s := by rw [← hs, Nat.Icc_insert_succ_left hmn]
+
+private lemma Icc_of_eq (hmn : m = n) : Icc m n = {m} := by simp [hmn]
+
+private lemma Ico_eq_of_Icc_pred_eq (hn : n ≠ 0) (hs : Icc m (n - 1) = s) : Ico m n = s := by
+  rw [← hs, Nat.Icc_pred_right _ hn.bot_lt]
+
+private lemma Ico_zero (m : ℕ) : Ico m 0 = ∅ := by simp
+
+private lemma Ioc_eq_of_Icc_succ_eq (hs : Icc (m + 1) n = s) : Ioc m n = s := by
+  rw [← hs, Nat.Icc_succ_left]
+
+private lemma Ioo_eq_of_Icc_succ_pred_eq (hs : Icc (m + 1) (n - 1) = s) : Ioo m n = s := by
+  rw [← hs, ← Nat.Icc_eq_Ioo]
+
+end lemmas
+
+/-- Given natural numbers `m` and `n`, returns `(s, ⊢ Finset.Icc m n = s)`. -/
+partial def evalFinsetIccNat {em eml en enl : Q(ℕ)} (m n : ℕ) (hm : Q(IsNat $em $eml))
+    (hn : Q(IsNat $en $enl)) : MetaM ((s : Q(Finset ℕ)) × Q(.Icc $em $en = $s)) := do
+  -- If `m = n`, then `Icc m n = {m}`. We handle this case separately because `insert m ∅` is
+  -- not synteq to `{m}`.
+  if m = n then
+    let hmn ← mkDecideProofQq q($em = $en)
```

Comment:
No need to `decide`, you can just use `Eq.refl` to get the equality of raw nat literals

Reply:
How? `let hmn : Q($em = $en) := q(rfl)` rightfully complains

Reply:
This should work:
```
have : $em =Q $en := \<\> 
```
then you don't even need to build the `rfl` proof

Reply:
Okay, I see. You're claiming that decide-eq numerals are actually defeq